gigascience/gigadb-website
# Upgrade codebase from Vue.js 2 to Vue.js 2.7

# Pull request for issue: #1624

This is a pull request for the following functionalities:

* Upgrade vue to latest Vue 2 version (2.7)
* Add way to watch and rebuild vue app on file changes

## How to test?

* When you switch to this branch, either re-run the `up.sh` script or if you run `docker-compose run --rm js bash -c "npm install ; npm run build ; npm run deploy"` it should work too (once you switch back to your working branch, if you're doing something related to the vue widget, you might want to rerun these again)
* Go over FUW flow and verify that nothing has changed
* Optionally install Vue dev tools for chrome or firefox, navigate to the file upload page, open the dev tools (will have to close and open dev tools after installing extension so it gets refreshed) and verify that the Vue version is 2.7
![image](https://github.com/gigascience/gigadb-website/assets/143437854/48765ef6-2990-43ee-b46e-e44e155eeb57)
  * Vue dev tools chrome https://chromewebstore.google.com/detail/vuejs-devtools/nhdogjmejiglipccpnnnanhbledajbpd
  * vue dev tools firefox https://addons.mozilla.org/en-US/firefox/addon/vue-js-devtools/

## How have functionalities been implemented?

* Vue version was upped to 2.7.0 in package.json, then it was installed with `npm install` and then app was rebuilt. I verified FUW functionality works as expected (no changes)
* Added one liner script that runs chokidar to watch for changes in the ./src folder and rebuild and serve widget
* Added `.nvmrc` file for node versioning based on the version in the file `.gitlab-ci.yml` (`17.6.0`)
  * nvm is a version manager for node adn npm that allows to easily switch node versions. An `.nvmrc` file keeps track of the expected version and switches to it with the command `nvm use`. Having this file there only matters if the dev uses `nvm` otherwise it doesn't affect anything else, it just a helper to set the node version in the current terminal.

NOTES:
* currently recommended node version is 20.11 wherease app uses 17.6. Could be a good idea to upgrade now along the rest of upgrades.
* Based on this, Vue 2 is EOL: https://v2.vuejs.org/lts/ (ending Dec. 31st, 2023 after which no more updates will happen), for more long term support (security, browser compatibility, ...) there is Vue 2 NES (never ending support) https://www.herodevs.com/support/nes-vue which is a paid service. An alternative is to update to Vue 3 eventually.

## Any issues with implementation?

--

## Any changes to automated tests?

--

## Any changes to documentation?

Added a README.md to the Vue root folder

## Any technical debt repayment?

Would be good to setup a better way to watch and serve changes in the Vue app, ideally with serving locally rather than building the app each time

## Any improvements to CI/CD pipeline?

--
